### PR TITLE
BHoM managed Python virtualenvs

### LINKED ISSUE #46: Python_Toolkit: InstallPython fails when force is set to true
#### Description:

If python is already installed in BHoM folder, the component failed with the error message saying "Coule not install Python".
Some investigation needed but my initial thought is some files may be used that is causing a fail of `ZipFile.ExtractToDirectory(pythonZip, pythonHome)`

#### How to replicate:

Install Python Toolkit
Reinstall with force set to true

#### Expected behaviour:

Python toolkit to be re-install

### LINKED ISSUE #62: Python_Toolkit: Execute python script
#### Description:

A generic execute function of a python script will be helpful for multiple toolkits.

### LINKED ISSUE #79: Python_Toolkit: Add 3.7.2, enable formatting, register downstream environments in ipykernel
#### Description:

A range of additional tweaks to better set-up the BHoM Python environments.
- In order to allow Python 3.7.2 to be installed for a downstream toolkit to make use of this particular version of Python.
- Add "black" to default packages to enable formatting of BHoM Python code across any toolkits
- Register downstream environments with root environment ipykernel to enable access via Jupyter Notebooks
- Ensure environment-specific code is accessible in notebooks by prepending the `./PythonCode` directory to PATH in the base kernel upon creation.

### LINKED ISSUE #81: InstallPython fails when Python already exists and `force`=true
#### How to replicate:
Have the following file already exist: 'C:\\ProgramData\\BHoM\\Extensions\\Python\\python.exe'. Then, run InstallPython with `force`=true. 

Fails on line 61, "The file 'C:\\ProgramData\\BHoM\\Extensions\\Python\\python.exe' already exists."

#### Expected behaviour:
Python successfully reinstalled, success == true.

### LINKED ISSUE #83: Simplify toolkit/workflows and enable reference of external libraries
#### Description:

Several tools we use come with their own Python versions/libraries. It would be nice to enable these to be used in BHoM workflows (with appropriate checks in place to ensure the external env is appropriate for the toolkit/process).

It would also be nice to do a bit of spring cleaning to remove some outdated functionality as we're moving towards an "environment-per-toolkit" ecosystem and being less constrictive with how those environments can be set-up. Plus once the changes mentioned here are implemented, we'd have 3 different ways of installing Python, which is silly.

Key features within this include:
- Creation of a BHoM Python Environment from an existing python.exe located anywhere on system
- Installation of a BHoM Python environment
- Addition of toolkit-specific code to environment site-packages in order to support their use within that environment
- Removal of toolkit-specific code from environment site-packages, for updates from BHoM workflows using external environments
- Installation of packages using pip from a requirements.txt file
- Checking of env for existence of packages within a requirements.txt file
- Removal of a BHoM Python environment

### PR DESCRIPTION
**Pull request by tg359**
### Issues addressed by this PR

Closes #83 
Closes #81 
Closes #79 
Closes #62 
Closes #46 

Handling of Python environments via BHoM rationalised around a base "full" environment, and virtualenvs for any toolkit/project-specific environment that can be pre-defined with custom codebases via a referenced local-package added upon installation.

### Test files

[83.zip](https://github.com/BHoM/Python_Toolkit/files/9336409/83.zip)

### CODE REVIEW

**Comment by tg359** [comment]
@BHoMBot check required

File: Python_Engine/Compute/DownloadPython.cs [hunk 1]
```diff
@@ -20,43 +20,45 @@
...
 namespace BH.Engine.Python
 {
     public static partial class Compute
     {
-        /***************************************************/
-        /**** Public Methods                            ****/
-        /***************************************************/
-
-        public static bool DownloadPython()
+        [Description("Download a the target version of Python.")]
```
Comment (alelom) [inline Python_Engine/Compute/DownloadPython.cs:32]:
```suggestion
        [Description("Download a target version of Python.")]
```

File: Python_Engine/Compute/InstallBasePythonEnvironment.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+{
+    public static partial class Compute
+    {
+        [Description("Install the base Python Environment for BHoM workflows.")]
+        [Input("run", "Run the installation process for the BHoM Python Environment.")]
+        [Output("env", "The base Python Environment for BHoM workflows.")]
+        public static oM.Python.PythonEnvironment InstallBasePythonEnvironment(bool run = false)
+        {
+            if (run)
+            {
```
Comment (alelom) [inline Python_Engine/Compute/InstallBasePythonEnvironment.cs:40]:
```suggestion
            if (!run)
            return
```

Comment (alelom) [inline Python_Engine/Compute/InstallBasePythonEnvironment.cs:40]: Please remove indented block
```suggestion
            if (!run)
                return;
```

File: Python_Engine/Compute/InstallReferencedVirtualenv.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            }
+
+            if (!File.Exists(executable))
+            {
+                BH.Engine.Base.Compute.RecordError($"{executable} doesn't exist. Have you installed LadybugTools Pollination?");
+                return null;
+            }
+
+            if (run)
+            {
```
Comment (alelom) [inline Python_Engine/Compute/InstallReferencedVirtualenv.cs:60]: Please remove indented block. Perhaps consider adding a warning to alert user to toggle `run` to `true`.
```suggestion
            if (!run)
            {
               BH.Engine.Base.Compute.RecordWarning($"Please toggle `{nameof(run)}` to true.");
               return;
            }
```

File: Python_Engine/Compute/InstallVirtualenv.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        {
+
+            if (!IsValidEnvironmentName(name))
+            {
+                BH.Engine.Base.Compute.RecordError("A BHoM Python virtualenv cannot cannot contain invalid filepath characters.");
+                return null;
+            }
+
+            if (run)
+            {
```
Comment (alelom) [inline Python_Engine/Compute/InstallVirtualenv.cs:55]: Please remove indented block. Perhaps consider adding a warning to alert user to toggle `run` to `true`.
```suggestion
            if (!run)
            {
               BH.Engine.Base.Compute.RecordWarning($"Please toggle `{nameof(run)}` to true.");
               return null;
            }
```

File: Python_Engine/Compute/RemoveVirtualenv.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        [Input("name", "The name of this virtualenv to remove.")]
+        [Input("run", "Run the installation process for this virtualenv.")]
+        [Output("success", "True if the environment of the given name has been removed.")]
+        public static bool RemoveVirtualenv(
+            string name,
+            bool run = false
+        )
+        {
+            if (run)
+            {
```
Comment (alelom) [inline Python_Engine/Compute/RemoveVirtualenv.cs:44]: Please remove indented block. Perhaps consider adding a warning to alert user to toggle `run` to `true`.
```suggestion
            if (!run)
            {
               BH.Engine.Base.Compute.RecordWarning($"Please toggle `{nameof(run)}` to true.");
               return null;
            }
```

File: Python_Engine/Compute/Run.cs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                process.StartInfo.WindowStyle = System.Diagnostics.ProcessWindowStyle.Hidden;
+                commandMode = "/C";
+            }
+            process.StartInfo.FileName = "cmd.exe";
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.UseShellExecute = false;
+            process.StartInfo.Arguments = $"{commandMode} {command}";
+            process.Start();
+
+            process.WaitForExit();
```
Comment (alelom) [inline Python_Engine/Compute/Run.cs:90]: What happens if the process hangs indefinitely without exiting? I think this way the process may deadlock Grasshopper's UI. Shouldn't we consider something like:
```suggestion
            int timeoutMinutes = 5; // should be an input parameter to the method
            int millisecondsToWait = 5 * 60 * 1000;
            process.WaitForExit(millisecondsToWait);
            if (!myProcess.HasExited)
		process.Kill();
```

File: Python_Engine/Compute/Run.cs [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.RedirectStandardError = true;
+            process.StartInfo.UseShellExecute = false;
+            process.StartInfo.Arguments = $"{commandMode} {command}";
+            process.Start();
+
+            // To avoid deadlocks, always read the output stream first and then wait.  
+            string stdout = process.StandardOutput.ReadToEnd();
+            string stderr = process.StandardError.ReadToEnd();
+            process.WaitForExit();
```
Comment (alelom) [inline Python_Engine/Compute/Run.cs:126]: What happens if the process hangs indefinitely without exiting? I think this way the process may deadlock Grasshopper's UI. Shouldn't we consider something like:
```suggestion
            int timeoutMinutes = 5; // should be an input parameter to the method
            int millisecondsToWait = 5 * 60 * 1000;
            process.WaitForExit(millisecondsToWait);
            if (!myProcess.HasExited)
		process.Kill();
```

**Review by alelom — CHANGES_REQUESTED** [review]
Looks great! I have not yet tested it, just had a look at the code. Will test it soon. Could you have a look at my suggestions?

**Review by alelom — DISMISSED** [review]
Thanks for the changes @tg359 , code now looks good to me.
The test file run as in the following image.  
It's not explicit whether the second component should fail intentionally, but it looks like so (incorrect python command requested).
Another note is on the runtime: on my machine, the `InstallBasePythonEnvironment` method took over 3 minutes (see screenshot), but I guess that is unavoidable. The other components took an average of 2 secs.
image

Approving based on the above.

**Comment by tg359** [comment]
@BHoMBot check required

**Comment by tg359** [comment]
@BHoMBot check compliance

**Comment by tg359** [comment]
@BHoMBot check compliance

**Comment by tg359** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by tg359** [comment]
@BHoMBot check versioning

**Comment by tg359** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling
@BHoMBot check core
@BHoMBot check serialisation

**Review by FraserGreenroyd — APPROVED** [review]
Approving based on @alelom review as commits since then have been versioning and compliance.

**Pull request merged**